### ISSUE TITLE
Time-based crossvalidation in AutoGluon Tabular?

### ISSUE THREAD

**Issue by schinto** [autogluon/autogluon#617]
Is [time-based-cross-validation](https://towardsdatascience.com/time-based-cross-validation-d259b13d42b8) supported in AutoGluon Tabular?

**Comment by Innixma** [comment]
Not at present, however this is an active area where we are investigating to expand AutoGluon's functionality to time-series problems.

As a clarification, this cross-validation technique would not produce predictions on all training rows since the earliest in time rows would never be part of a validation set during cross-val, is that correct? Thus, with 5-fold cross-val only 4 folds (80%) of the data will receive out-of-fold predictions. That is the primary difficulty, as with multi-layer stacking, future layers would have to remove this data.

**Comment by kmedved** [comment]
That's correct, yes @Innixma - the earliest in time rows would never be included in the validation data. Depending on the size of your training data, you can do 10 folds to ensure that 90% of your data receives a prediction, etc..., but to my knowledge there's no way around the core issue you raise.

**Comment by Innixma** [comment]
As an update to this feature:

We are currently working on time-series functionality. The PR of the WIP time-series support is available here: #684

At present, it's recommended to not use bagging / stacking for time series problems in AutoGluon and instead use a train/val split and pass both into fit as `train_data`, `tuning_data`. Once #684 is merged, the plan is to investigate adding the inner cross-val time-series support for time-series ensembling.

Typically, this technique is used for HPO, so it's unclear if it will give a performance boost in the same way we get a boost in traditional supervised learning, as the amount of training data at each layer will shrink. With enough folds I'd bet it will still be an improvement. Either way, we are interested to try it out.

**Comment by schinto** [comment]
A generalized form of the Splitter concept for CV as used here would be great
https://github.com/deepchem/deepchem/blob/master/deepchem/splits/splitters.py
So that time/clusterid/integer columns can be selected, which contains custom splits, but the columns are excluded from prediction.

**Comment by jwmueller** [comment]
@Innixma time-based crossvalidation or train/val split does not necessarily correspond to time-series. It may still be useful for regular supervised learning where rows are independent, just not identically distributed over time.

I think users who want a time-based single train/val split should just do it themselves though and pass their val data as `tuning_data`.  For bagging/stacking, we could consider supporting arbitrary user-defined k-fold split, although this still doesn't resolve issue of being not fully compatible with a time-based crossvalidation schema due to the issue you mentioned.

Maybe we should provide some advice for handling time column in our Tabular FAQ section? I'm not sure what exactly to recommend though

**Comment by Innixma** [comment]
@jwmueller my current plan is to add generic support for specifying fold indices (although it will disable repeated bagging when specified). The time-based cross-val will be a separate implementation, since it needs special handling.

**Comment by jeremyhermann** [comment]
Hello - have you made any progress on temporal (or fully custom) cv splits that also supports bagging/stacking? Thanks

**Comment by Innixma** [comment]
Hi @jeremyhermann , we are currently working on adding AutoGluon-Forecasting to v0.5 which handles temporal data. As part of this we are working on support for temporal CV splits for tuning models, although this will not be used in bagging/stacking in v0.5, that requires additional work that may be pursued in future.

**Comment by SamuelDietz** [comment]
Hey there -

any updates on this topic? I'm really looking forward to it (or to be able to use unknown covariates in time series forecasting).

As I really like AutoGluon I would love to use if for sophisticated time series forecasting but for that I need the above mentioned I think. The milestone modifications are scaring me a bit 😉

Thank you!

**Comment by canerturkmen** [comment]
Hi @SamuelDietz thank you for your question. 

AutoGluon-TimeSeries went out with v0.5, and multi-window backtesting capability was added with v0.6. see https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-indepth.html#multi-window-backtesting

We are looking forward to adding past-only covariates in v0.7. 

Hope this helps,
Caner

**Comment by SamuelDietz** [comment]
> We are looking forward to adding past-only covariates in v0.7.

Awesome, looking forward to it, thanks!

**Comment by canerturkmen** [comment]
Hi @SamuelDietz, Past only covariate support was also added with #2680 . Please let us know if this helps and your experience with it.  Closing this for now.
